### PR TITLE
feat: use LRU caches for proposer and execution caches

### PR DESCRIPTION
**Pull request by 0xYYY**
# Motivation
Closes https://github.com/ralexstokes/relay-monitor/issues/5.

# Solution
Replace proposer and execution caches with LRU caches using [golang-lru](https://github.com/hashicorp/golang-lru).

# Note
I have little to no experience in Go so please let me know if I'm doing anything incorrectly.

### CODE REVIEW

File: pkg/consensus/client.go [hunk 1]
```diff
@@ -49,36 +50,47 @@ func NewClient(endpoint string, clock *Clock, logger *zap.Logger) *Client {
...
+	validator, ok := c.proposerCache.Get(slot)
 	if !ok {
 		return nil, fmt.Errorf("missing proposal for slot %d", slot)
 	}
 
-	return &validator.publicKey, nil
+	if v, ok := validator.(ValidatorInfo); ok {
+		return &v.publicKey, nil
+	} else {
+		return &types.PublicKey{}, fmt.Errorf("invalid value stored in cache")
```
Comment (terencechain) [inline pkg/consensus/client.go:92]:
```suggestion
		return &types.PublicKey{}, errors.New("invalid value stored in cache")
```

Even better, store it in variable and reuse it for `fetchExecutionHash`

Comment (0xYYY) [reply pkg/consensus/client.go]: thanks for the suggestion, added a error value for this

**Comment by terencechain** [comment]
Hi @0xYYY, do you mind fixing the conflicts? Thanks!

**Comment by 0xYYY** [comment]
@terencechain done!

**Review by terencechain — APPROVED** [review]
Thanks for your contribution! I'll be happy to add unit tests after

**Comment by 0xYYY** [comment]
fixed another merge conflict, please let me know if there  are anything else required for this PR
would be happy to contribute on unit tests in a follow up PR

**Comment by ralexstokes** [comment]
@0xYYY thanks for this!

I'd support adding LRU caches like you have done to the current `proposerCache` and `executionCache` fields

there has been a bit of refactoring over the last few days so I'd suggest just taking solely the LRU changes from this PR and applying them in a fresh branch

thanks!

**Comment by 0xYYY** [comment]
sure thing! i'll close this PR and submit a new one to replace it

**Pull request closed without merging**